bee-ledger/src/types/receipt.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::{error::Error, TreasuryOutput};
+
+use bee_common::packable::{Packable, Read, Write};
+use bee_message::{
+    constants::IOTA_SUPPLY,
+    input::Input,
+    milestone::MilestoneIndex,
+    output::Output,
+    payload::{receipt::ReceiptPayload, Payload},
+};
+
+/// A type that wraps a receipt and the index of the milestone in which it was included.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Receipt {
+    inner: ReceiptPayload,
+    included_in: MilestoneIndex,
+}
+
+impl Receipt {
+    /// Creates a new `Receipt`.
+    pub fn new(inner: ReceiptPayload, included_in: MilestoneIndex) -> Self {
+        Self { inner, included_in }
+    }
+
+    /// Returns the inner receipt of the `Receipt`.
+    pub fn inner(&self) -> &ReceiptPayload {
+        &self.inner
+    }
+
+    /// Returns the index of the milestone in which the `Receipt` was included.
+    pub fn included_in(&self) -> &MilestoneIndex {
+        &self.included_in
+    }
+
+    /// Semantically validates the `Receipt`.
+    pub fn validate(&self, consumed_treasury_output: &TreasuryOutput) -> Result<(), Error> {
+        let mut migrated_amount: u64 = 0;
+        let transaction = match self.inner().transaction() {
+            Payload::TreasuryTransaction(transaction) => transaction,
+            payload => return Err(Error::UnsupportedPayloadKind(payload.kind())),
+        };
+
+        for funds in self.inner().funds() {
+            migrated_amount = migrated_amount.checked_add(funds.output().amount()).ok_or_else(|| {
+                Error::InvalidMigratedFundsAmount(migrated_amount as u128 + funds.output().amount() as u128)
+            })?;
+        }
+
+        if migrated_amount > IOTA_SUPPLY {
+            return Err(Error::InvalidMigratedFundsAmount(migrated_amount as u128));
+        }
+
+        match transaction.input() {
+            Input::Treasury(input) => {
+                if input.milestone_id() != consumed_treasury_output.milestone_id() {
+                    return Err(Error::ConsumedTreasuryOutputMismatch(
+                        *input.milestone_id(),
+                        *consumed_treasury_output.milestone_id(),
+                    ));
+                }
+            }
+            input => return Err(Error::UnsupportedInputKind(input.kind())),
+        };
+
+        let created_treasury_output = match transaction.output() {
+            Output::Treasury(output) => output,
+            output => return Err(Error::UnsupportedOutputKind(output.kind())),
+        };
+
+        let created_amount = consumed_treasury_output
+            .inner()
+            .amount()
+            .checked_sub(migrated_amount)
+            .ok_or_else(|| {
+                Error::InvalidMigratedFundsAmount(
+                    consumed_treasury_output.inner().amount() as u128 - migrated_amount as u128,
```

Comment:
this will also underflow for `u128`. Did you meant to write `i128`?

Reply:
Good catch